File: python/infinicore/nn/modules/linear.py
import torch
import infinicore
from typing import Union
from .module import Module

class Linear(Module):
    r"""Applies an affine linear transformation to the incoming data: :math:`y = xA^T + b`.

    Args:
        in_features: size of each input sample
        out_features: size of each output sample
        bias: If set to ``False``, the layer will not learn an additive bias.
            Default: ``True``

    Shape:
        - Input: :math:`(*, H_\text{in})` where :math:`*` means any number of
          dimensions including none and :math:`H_\text{in} = \text{in\_features}`.
        - Output: :math:`(*, H_\text{out})` where all but the last dimension
          are the same shape as the input and :math:`H_\text{out} = \text{out\_features}`.

    Attributes:
        weight: the learnable weights of the module of shape
            :math:`(\text{out\_features}, \text{in\_features})`. The values are
            initialized from :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})`, where
            :math:`k = \frac{1}{\text{in\_features}}`
        bias:   the learnable bias of the module of shape :math:`(\text{out\_features})`.
                If :attr:`bias` is ``True``, the values are initialized from
                :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})` where
                :math:`k = \frac{1}{\text{in\_features}}`
    """

    __constants__ = ["in_features", "out_features"]
    in_features: int
    out_features: int
    weight: infinicore.Tensor

    def __init__(
        self,
        in_features: int,
        out_features: int,
        bias: bool = False,
        device=None,
        dtype=None,
    ) -> None:
        factory_kwargs = {"device": device, "dtype": dtype}
        super().__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.weight = infinicore.nn.Parameter(
            infinicore.empty((1,), dtype=infinicore.float32, device=infinicore.device("cpu", 0))
        )
    
    def forward(self, input: infinicore.Tensor) ->  infinicore.Tensor:
        return infinicore.nn.functional.linear(input, self.weight, None)
    
    def extra_repr(self) -> str:
        return f"in_features={self.in_features}, out_features={self.out_features}"

File: src/infinicore/ops/linear/linear.cc
#include "infinicore/ops/linear.hpp"
#include "infinicore/ops/add.hpp"
#include "infinicore/ops/matmul.hpp"
#include "infinicore/ops/rearrange.hpp"

namespace infinicore::op {

void print(Shape &shape) {
    printf("\n");
    for (auto &v : shape) {
        printf("\t %ld", v);
    }
    printf("\n");
}

Tensor linear2(Tensor input,
               Tensor weight,
               pybind11::object bias) {

    Size ndim = input->ndim();
    Shape input_shape = input->shape();
    Shape weight_shape = weight->shape();

    Size num = input_shape[0];
    Size in_features = input_shape[ndim - 1];
    Size out_features = weight_shape[1];

    // y 是 (∗,out_features)
    auto y = Tensor::empty({num, out_features}, input->dtype(), input->device());

    matmul_(y, input, weight);

    if (!bias.is_none()) {
        Tensor bias_tensor = bias.cast<Tensor>();
        add_(y, y, bias_tensor);
    }
    return y;
}

Tensor linear(Tensor input,  // (∗,in_features) where * means any number of additional dimensions, including none
              Tensor weight) //(out_features, in_features)
{
    Size ndim = input->ndim();
    Shape input_shape = input->shape();
    Shape weight_shape = weight->shape();

    Size num = input_shape[0];
    Size in_features = input_shape[ndim - 1];
    Size out_features = weight_shape[1];

    // y 是 (∗,out_features)
    auto y = Tensor::empty({num, out_features}, DataType::F64, input->device());
    y = Tensor::empty({num, out_features}, input->dtype(), input->device());

    matmul_(y, input, rearrange(weight));
    return y;
}

Tensor linear_bk(Tensor input,
                 Tensor weight) {
    Size ndim = input->ndim();
    Shape input_shape = input->shape();
    Shape weight_shape = weight->shape();

    Size num = input_shape[0];
    Size in_features = input_shape[ndim - 1];
    Size out_features = weight_shape[1];

    // y 是 (∗,out_features)
    auto y = Tensor::empty({num, out_features}, input->dtype(), input->device());
    matmul_(y, input, rearrange(weight));
    return y;
}

Tensor linear_bias(Tensor input,  //  (∗,in_features)
                   Tensor weight, // (in_features,out_features)
                   Tensor bias    // (out_features)
) {
    Tensor y = linear(input, weight);
    add_(y, y, bias);
    return y;
}

} // namespace infinicore::op
